library(dplyr)
library(data.table)
library(TwoSampleMR)
library(optparse)

# Load arguments ----------------------------------------------------------
# Environment: R 3.6.1

parse <- OptionParser()

option_list <- list(
   make_option('--im', type='character', help="Imaging GWAS files in rds format", action='store'),
   make_option('--interexp', type='character', 
               help='Folder for intermediate files generated by processing exposure data', action='store')
)

args = commandArgs(trailingOnly=TRUE)

opt <- parse_args(OptionParser(option_list=option_list), args=args)
exposure.path=opt$im
output.path=opt$interexp

# Basic settings ----------------------------------------------------------

dir.create(file.path(output.path), showWarnings = FALSE)

# Process exposure data ---------------------------------------------------
ls.exposure='MDD'
ls.exposure.filename = list.files(path = exposure.path, pattern = '.forPRSice',full.names=T) 

for (f in 1:length(ls.exposure)){
      
      summ.stats.fname=ls.exposure.filename[f]
      gwas_summarystat = fread(summ.stats.fname,header=T,stringsAsFactors=F) %>% 
         select(SNP,A1,A2,BETA,SE,P,Direction,af=Freq1,CHR,BP) %>%
         dplyr::mutate(N=807553)
      traitA=ls.exposure[f]
      
      ### find top SNPs
      
      colnames(gwas_summarystat) = c('SNP','Allele1','Allele2','Effect','se','p','Direction','Freq1','CHR','BP','N')

      SigniSNP=filter(gwas_summarystat,p<5e-8)
      
      SigniSNP$NewAllele1=SigniSNP$Allele1
      SigniSNP$NewAllele2=SigniSNP$Allele2
      SigniSNP$NewFreq1=SigniSNP$Freq1
      SigniSNP$NewEffect=SigniSNP$Effect
      
      lst=c(which(SigniSNP$Effect<0))
      SigniSNP$NewEffect[lst]=-1*SigniSNP$Effect[lst]
      SigniSNP$NewFreq1[lst]=1-1*SigniSNP$Freq1[lst]
      SigniSNP$NewAllele1[lst]=SigniSNP$Allele2[lst]
      SigniSNP$NewAllele2[lst]=SigniSNP$Allele1[lst]
      SigniSNP=data.frame(SigniSNP)
      SigniSNP=SigniSNP[,c('SNP','NewAllele1','NewAllele2','NewFreq1','NewEffect','se','p','N')]
      
      colnames(SigniSNP)=c("SNP","effect_allele","other_allele","eaf", "beta","se","pval",'samplesize')
      SigniSNP=data.frame(Phenotype=traitA,units='unit',SigniSNP)
      
      write.table(SigniSNP,file=paste0(output.path,'/',traitA,".tops"),col.names=T,row.names = F,sep="\t",quot=F)
      
      
      ### clump exposure dat
      
      
      #read broad as exposure
      exposure_dat <- read_exposure_data(paste0(output.path,'/',traitA,".tops"),sep='\t')
      #clump
      exposure_dat <- clump_data(exposure_dat,clump_kb=250,clump_r2=0.001)
      #exposure_dat <- filter(exposure_dat,eaf.exposure>0.1,eaf.exposure<0.90,abs(beta.exposure)<0.038)
      exposure_dat <- exposure_dat[,c('SNP','effect_allele.exposure','other_allele.exposure',
                                      'eaf.exposure','beta.exposure','se.exposure','pval.exposure','samplesize.exposure',
                                      'units.exposure','exposure','mr_keep.exposure','pval_origin.exposure',
                                      'units.exposure_dat','id.exposure','data_source.exposure')]
      write.table(exposure_dat,file=paste0(output.path,'/',traitA,".exposure_dat"),col.names=T,row.names = F,sep="\t",quote=F)
      
      system(paste0('echo ',grep(f,ls.exposure),' ',traitA,'   exposure prep done >> XS.log'))
      #system(paste0('wc -l ./exposure_stats/',traitA,'.exposure_dat >> XS.log'))
      
}

rm(list=ls())